### ISSUE TITLE
Help needed updating from 4.x.x to 6.2.0

### ISSUE THREAD

**Issue by maxwellhadley** [node-serialport/node-serialport#1578]
- __SerialPort Version__: 6.2.0
- __NodeJS Version__: v6.11.5
- __Operating System__ and __Hardware Platform__: MacOS High Sierra 10.13.15, iMac (21.5-inch, Late 2009) 

- Have you checked the right __version__ of the __api docs?__: Yes
- Are you having trouble installing and you checked the __Installation Special Cases__ docs? N/A
- Are you using __Electron__ and have you checked the _Electron Docs_?: N/A

### Summary of Problem
I maintain the package **node-rfxcom** [https://github.com/rfxcom/node-rfxcom](url) which uses serialport to talk to an RFXCOM RFXtrx433E transceiver (internally this uses and FTDI device). The current version of node-rfxcom (1.3.1) uses serialport 4.x.x. This works, and specifically it handles the device being unplugged and plugged in again at random, reconnecting automatically.

I am trying to update to serialport 6.2.0 and I cannot obtain reliable handling of plug/unplug events. As far as I can tell, if the RFXtrx is once unplugged when connected to serialport, the only way to restore reliable operation is to reboot the computer. Otherwise a wide range of weird things happen, including being able to read but not write, or receiving extra data events apparently at random, which look like a modified copy of previously received data. I have not been able to come up with a simple sequence of of steps that will reliably reproduce a consistent error behaviour.

Basically I'm looking for help: there might be an issue with serialport, or I might just not have understood the changes between 4.x.x and 6.2.0

### Steps and Code to Reproduce the Issue

1. Purchase an RFXCOM RFXtrx433E [http://www.rfxcom.com/epages/78165469.sf/en_GB/?ObjectPath=/Shops/78165469/Products/14103](url)

2. Fork [https://github.com/rfxcom/node-rfxcom](url) and check out the use-serialport-6 branch

3. The README contains sample code to create and use an rfxcom object, and the documentation describes how to use the various events to handle connection & disconnection

**Comment by maxwellhadley** [comment]
I eventually solved the problem. There were two issues:

- Writes to a closed serialport are now buffered internally and will be sent if & when the port is subsequently (re)opened
- If a write is 'in progress' when the device is disconnected, you get an`error` event **as well as** the expected `close` event.

Re-writing the disconnect logic to account for these gives me a working system. However, there is still a race condition. I am guarding all writes by first testing if the port is open: if the device is disconnected between the access to the isOpen property and the write, the write will be queued when I don't want it to be.
